File: src/main/java/at/yawk/fiction/impl/fanfiction/TagParser.java
package at.yawk.fiction.impl.fanfiction;

import at.yawk.fiction.Chapter;
import at.yawk.fiction.impl.PageParser;
import com.google.common.collect.Lists;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.joda.time.Instant;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

/**
 * @author yawkat
 */
class TagParser extends PageParser<FfnStory> {
    @Override
    protected FfnStory create() {
        return new FfnStory();
    }

    @Override
    protected void parse(Element root, FfnStory target) throws Exception {
        Elements timeElements = root.select("span");
        List<String> items = split(root.text(), " - ");

        target.setGenres(Lists.transform(Arrays.asList(items.remove(2).split("/")), name -> {
            FfnGenre genre = new FfnGenre();
            genre.setName(name);
            return genre;
        }));

        target.setFavorites(parseIntLenient(removeMatch(items, "Favs: (\\d+)")));
        target.setFollows(parseIntLenient(removeMatch(items, "Follows: (\\d+)")));
        target.setWords(parseIntLenient(removeMatch(items, "Words: (\\d+)")));

        removeMatch(items, "Updated: (.*)");
        removeMatch(items, "Published: (.*)");
        target.setUpdateTime(new Instant(Long.parseLong(timeElements.first().attr("data-xutime")) * 1000));
        target.setPublishTime(new Instant(Long.parseLong(timeElements.last().attr("data-xutime")) * 1000));

        int chapterCount = Integer.parseInt(removeMatch(items, "Chapters: (\\d+)"));
        List<Chapter> chapters = new ArrayList<>(chapterCount);
        for (int i = 0; i < chapterCount; i++) {
            FfnChapter chapter = new FfnChapter();
            chapter.setIndex(i);
            chapters.add(chapter);
        }
        target.setChapters(chapters);
    }
}

File: src/main/kotlin/at/yawk/fiction/impl/fanfiction/TagParser.kt
package at.yawk.fiction.impl.fanfiction

import at.yawk.fiction.Chapter
import at.yawk.fiction.Story
import at.yawk.fiction.impl.PageParser
import org.joda.time.Instant
import org.jsoup.nodes.Element
import java.util.*

/**
 * @author yawkat
 */
internal object TagParser : PageParser<Story>() {
    override fun parse(root: Element): Story {
        val timeElements = root.select("span")
        val items = split(root.text(), " - ")

        val genres: List<FfnGenre>
        if (!items[2].startsWith("Chapters:")) {
            genres = items.removeAt(2).split("/").map { FfnGenre(name = it) }
        } else {
            genres = emptyList()
        }

        val favorites = parseIntLenient(removeMatch(items, "Favs: ([\\d,]+)"))
        val follows = parseIntLenient(removeMatch(items, "Follows: ([\\d,]+)"))
        val words = parseIntLenient(removeMatch(items, "Words: ([\\d,]+)"))

        removeMatch(items, "Updated: (.*)")
        removeMatch(items, "Published: (.*)")
        val updateTime = Instant(timeElements.first().attr("data-xutime").toLong() * 1000)
        val publishTime = Instant(timeElements.last().attr("data-xutime").toLong() * 1000)

        val chapterCount = parseIntLenient(removeMatch(items, "Chapters: ([\\d,]+)"))
        val chapters = ArrayList<Chapter>(chapterCount)
        for (i in 0..chapterCount - 1) {
            chapters.add(Chapter(providerData = FfnChapterProviderData(index = i)))
        }
        return Story(
                updateTime = updateTime,
                publishTime = publishTime,
                chapters = chapters,
                providerData = FfnStoryProviderData(
                        genres = genres,
                        favorites = favorites,
                        follows = follows,
                        words = words
                )
        )
    }
}
